File: Peare/Resources/RT_ICON_GROUP/RT_GROUP_ICON.cs
using System;
using System.Text;

namespace Peare
{
    public class RT_GROUP_ICON
    {
        public static string Get(byte[] data)
        {
            if (data == null || data.Length < 6)
                return "Invalid data";

            StringBuilder sb = new StringBuilder();

            // ICONDIR
            ushort idReserved = BitConverter.ToUInt16(data, 0); // 0x00
            ushort idType = BitConverter.ToUInt16(data, 2);     // 0x01
            ushort idCount = BitConverter.ToUInt16(data, 4);    // 0x02

            sb.AppendLine($"RT_GROUP_ICON");
            sb.AppendLine("{");
            sb.AppendLine($"\tReserved: {idReserved}");
            sb.AppendLine($"\tType: {idType} (1 = Icon)");
            sb.AppendLine($"\tCount: {idCount}");

            int offset = 6;

            for (int i = 0; i < idCount; i++)
            {
                if (offset + 14 > data.Length)
                {
                    sb.AppendLine($"  [!] GRPICONDIRENTRY {i} incomplete.");
                    break;
                }

                byte bWidth = data[offset];
                byte bHeight = data[offset + 1];
                byte bColorCount = data[offset + 2];
                byte bReserved = data[offset + 3];
                ushort wPlanes = BitConverter.ToUInt16(data, offset + 4);
                ushort wBitCount = BitConverter.ToUInt16(data, offset + 6);
                uint dwBytesInRes = BitConverter.ToUInt32(data, offset + 8);
                ushort nID = BitConverter.ToUInt16(data, offset + 12);

                sb.AppendLine($"\tRT_ICON #{nID}");
                sb.AppendLine("\t{");
                sb.AppendLine($"\t\tSize: {bWidth}x{bHeight} px");
                sb.AppendLine($"\t\tColors: {(bColorCount == 0 ? ">8bpp" : bColorCount.ToString())}");
                sb.AppendLine($"\t\tPlanes: {wPlanes}");
                sb.AppendLine($"\t\tBitCount: {wBitCount}");
                sb.AppendLine($"\t\tBytes in Resource: {dwBytesInRes}");
                sb.AppendLine("\t}");

                offset += 14;
            }

            sb.AppendLine("}");

            return sb.ToString();
        }
    }
}

File: src/decoder/resources/RT_GROUP_ICON/RT_GROUP_ICON.cpp
#include "RT_GROUP_ICON.h"

#include "../RT_ICON/RT_ICON.h"

#include <QDebug>
#include <QStringList>
#include <QtEndian>

namespace peare {
namespace resources {
namespace {

quint16 ReadUInt16(const QByteArray& data, int offset)
{
    if (offset < 0 || offset + 2 > data.size())
        return 0;
    return qFromLittleEndian<quint16>(reinterpret_cast<const uchar*>(data.constData() + offset));
}

quint32 ReadUInt32(const QByteArray& data, int offset)
{
    if (offset < 0 || offset + 4 > data.size())
        return 0;
    return qFromLittleEndian<quint32>(reinterpret_cast<const uchar*>(data.constData() + offset));
}

} // namespace

ResourcePreview RT_GROUP_ICON::preview(const ResourceEntry& entry, const IResourceResolver& resolver)
{
    ResourcePreview preview;

    if (entry.data.size() < 6)
    {
        preview.error = QStringLiteral("Invalid data");
        return preview;
    }

    QStringList lines;

    // ICONDIR
    const quint16 idReserved = ReadUInt16(entry.data, 0); // 0x00
    const quint16 idType = ReadUInt16(entry.data, 2);     // 0x01
    const quint16 idCount = ReadUInt16(entry.data, 4);    // 0x02

    lines.append(QStringLiteral("RT_GROUP_ICON"));
    lines.append(QStringLiteral("{"));
    lines.append(QStringLiteral("\tReserved: %1").arg(idReserved));
    lines.append(QStringLiteral("\tType: %1 (1 = Icon)").arg(idType));
    lines.append(QStringLiteral("\tCount: %1").arg(idCount));

    int offset = 6;

    for (int i = 0; i < idCount; ++i)
    {
        if (offset + 14 > entry.data.size())
        {
            lines.append(QStringLiteral("  [!] GRPICONDIRENTRY %1 incomplete.").arg(i));
            break;
        }

        const quint8 bWidth = static_cast<quint8>(entry.data.at(offset));
        const quint8 bHeight = static_cast<quint8>(entry.data.at(offset + 1));
        const quint8 bColorCount = static_cast<quint8>(entry.data.at(offset + 2));
        const quint8 bReserved = static_cast<quint8>(entry.data.at(offset + 3));
        Q_UNUSED(bReserved);
        const quint16 wPlanes = ReadUInt16(entry.data, offset + 4);
        const quint16 wBitCount = ReadUInt16(entry.data, offset + 6);
        const quint32 dwBytesInRes = ReadUInt32(entry.data, offset + 8);
        const quint16 nID = ReadUInt16(entry.data, offset + 12);

        lines.append(QStringLiteral("\tRT_ICON #%1").arg(nID));
        lines.append(QStringLiteral("\t{"));
        lines.append(QStringLiteral("\t\tSize: %1x%2 px").arg(bWidth).arg(bHeight));
        lines.append(QStringLiteral("\t\tColors: %1")
                         .arg(bColorCount == 0 ? QStringLiteral(">8bpp") : QString::number(bColorCount)));
        lines.append(QStringLiteral("\t\tPlanes: %1").arg(wPlanes));
        lines.append(QStringLiteral("\t\tBitCount: %1").arg(wBitCount));
        lines.append(QStringLiteral("\t\tBytes in Resource: %1").arg(dwBytesInRes));
        lines.append(QStringLiteral("\t}"));

        if (const ResourceEntry* child = resolver.find(
                QStringLiteral("RT_ICON"),
                QStringLiteral("#%1").arg(nID),
                entry.language))
        {
            preview.images += RT_ICON::decode(child->data);
        }

        offset += 14;
    }

    lines.append(QStringLiteral("}"));
    preview.text = lines.join(QStringLiteral("\r\n"));
    return preview;
}

} // namespace resources
} // namespace peare
